### DISCUSSION TITLE
Java Classpath

### DISCUSSION THREAD

**Discussion by tripower** [MarkusJx/node-java-bridge#37] [Q&A] [answered]
Hello

another question 
in the "old" / other Java NodeJS Bridge it was  possible to set a whole path instead of single jars
is that here also possible and how?

old code => java.classpath.pushDir(localservicePath);

greetings
Klaus

**Comment by MarkusJx** [comment, marked as answer]
Uhm, no, that's currently not possible. A valid workaround would be to use `glob`:
```ts
import glob from 'glob';
import java from 'java-bridge';

java.classpath.append(glob.sync('*.jar', { cwd: 'your-dir-here' }));
```

But I'll add this to my list of things to do: #38

**Reply by tripower** [reply]
ok, hope this workaround works, because 180 files with the other way is ... :)
and that can also be done for other files? like dll, ...

import glob from 'glob';
import java from 'java-bridge';

java.classpath.append(glob.sync('*.jar', { cwd: 'your-dir-here' }));
java.classpath.append(glob.sync('*.dll', { cwd: 'your-dir-here' }));

**Reply by MarkusJx** [reply]
You *should* be able to pass anything that is allowed by java's own `java.net.URLClassLoader`, so...probably, yes

**Reply by tripower** [reply]
perfect like your support!

**Reply by tripower** [reply]
Hello

glob seems not to be working or
the result of glob raise an error at the java bridge code

next try with set classpath works not
Classnotfound Exception

next try with append of only one jar
seems to work better but other Exception
hope that will be solved if all jars appended

Error 1
![Screenshot_20221122_120329](https://user-images.githubusercontent.com/25051422/203300116-47168922-b8ca-4580-91e6-2d283f271be4.jpg)

Error 2
![Screenshot_20221122_120408](https://user-images.githubusercontent.com/25051422/203300167-0f191383-617e-4fc6-815e-2fb262784cad.jpg)

Error 3
![Screenshot_20221122_120448](https://user-images.githubusercontent.com/25051422/203300240-160058ab-7718-42fb-b76a-9a9f4fc980da.jpg)

**Reply by tripower** [reply]
@MarkusJx and on an other PC a get this error 
at 
const System = importClass('java.lang.System');

![error](https://user-images.githubusercontent.com/25051422/203350737-bd981bd6-04e7-4fd5-828f-b814f6dee7eb.jpg)

**Reply by MarkusJx** [reply]
This last one is probably caused by the library being compiled for the wrong target architecture, i.e. you are either trying to load a x64 binary on a x32 system or having a mismatch between your java and java-bridge architectures

**Reply by tripower** [reply]
ok, good to know
for the other pc I have written a directory file reader to make my own array for append
hope that works instead of glob
and no problems of Spring will be occur

**Reply by tripower** [reply]
@MarkusJx 
here is the result of the import of the 176 files
which crashes with my code like with glob
any ideas?

[error files.txt](https://github.com/MarkusJx/node-java-bridge/files/10069191/error.files.txt)

would you also be by DM avaiable?
maybe we can get in contact by XING (https://www.xing.com/profile/Klaus_Becker29)

**Reply by MarkusJx** [reply]
This is probably related to the array not being passed properly, I'll look into it (#39). Alternatively, you could also call `java.classpath.append` multiple times meanwhile:
```ts
for (const file of glob.sync('*.jar', { cwd: 'your-dir-here' })) {
    java.classpath.append(file);
}
```
This is a tad slower, but should work as a small workaround. A lot of workarounds, I know, I'm working on it.

Unfortunately, I'm not on Xing (or any similar platform [that's a lie, I'm on linkedin, but that profile is private and has never been used]) but maybe I could give you an e-mail address to message to, once that's working.

**Reply by tripower** [reply]
thx

slow is no problem 😂🐌
it should only work

It's a shame that github has no DM option
because everything must no be public 

thanks for code tip
hope test will work tomorrow

nice evening ⭐

**Reply by tripower** [reply]
hello @MarkusJx 

the loop with append seems to work but after that errors rules the world :)

at this call
const JSysInfo = importClass('at.oebb.ticketshop.localservice.services.SystemInfoService');
[fehler_bridge.txt](https://github.com/MarkusJx/node-java-bridge/files/10074831/fehler_bridge.txt)

I think it is like the issue with Spring

greetings Klaus

**Reply by MarkusJx** [reply]
This looks like you are missing a dependency, you can check your loaded jar files (or to be more precise, your class path) by calling:
```ts
import java from 'java-bridge';

console.log(java.classpath.get());
```

Also, if you need to send me anything without having to disclose any information publicly, you can send me an [E-Mail](mailto:hi@markusjx.com). Otherwise I'd just stick to these GitHub discussions, so others may find useful information within these answers.

**Reply by tripower** [reply]
normaly all files are appended
worked so in the old bridge

I just wondering that this happens at the import
because with one jar we have been at the next step

thx for your e-mail

greetings Klaus

**Reply by tripower** [reply]
hope more infos about the classpath will I get tomorrow
but still surprised that at the import call a dependency error can occur

**Reply by tripower** [reply]
@MarkusJx 
so
get shows that all my jars in the classpath
so no idea why the error occurs

would be the calls of my code
with usage of the other java bridge be helpful?
or is your implementation completely different?

**Reply by tripower** [reply]
@MarkusJx 
here attached an example code of the call with the "old" nodejs java bridge
same jars, same java, .... as now
worked with Electron v18 (but this should not here the problem)

[nodejs java bridge (old) - ls demo code.txt](https://github.com/MarkusJx/node-java-bridge/files/10085522/nodejs.java.bridge.old.-.ls.demo.code.txt)

**Reply by MarkusJx** [reply]
I mean, that *should* work. I'll try creating test cases for this specific use case, but creating the java instance with options already works, if adding your directory to the java library path:
```ts
import java from 'java-bridge';

java.ensureJvm({
    opts: [
        `-Djava.library.path=${localservicejarsPath};`,
        '-Dspring.application.admin.enabled=true;'
    ]
});
```

**Reply by tripower** [reply]
Hello @MarkusJx 

the lib path was only for DLLs which a printer need

same problems / error as before
and the "missing" class from the error is in spring-core-4.3.3.RELEASE.jar available
so no idea whats wrong

Code
[main.txt](https://github.com/MarkusJx/node-java-bridge/files/10086409/main.txt)
Output
[output.txt](https://github.com/MarkusJx/node-java-bridge/files/10086406/output.txt)
Error
[error.txt](https://github.com/MarkusJx/node-java-bridge/files/10086405/error.txt)

@MarkusJx maybe you could also try something with spring franework
for your information
my project is used as normal spring boot and also in this case like IoC Container 
https://www.geeksforgeeks.org/spring-ioc-container/
https://www.baeldung.com/inversion-control-and-dependency-injection-in-spring

**Reply by MarkusJx** [reply]
Just a quick update: I'm currently dealing with the application crashing on win32 (again), which was fixed previously, but something has re-introduced this issue on my current branch. So this may take a while...

**Reply by tripower** [reply]
@MarkusJx 
thx for the information
hope s solution can be found soon

greetings Klaus

**Reply by MarkusJx** [reply]
So this seems like a good old Heisenbug, which has been introduced by me doing nothing, which is great, so this may actually cost me a lot of time to figure out

**Reply by tripower** [reply]
@MarkusJx 
I have made some tests with the **other / old nodejs java bridge**
it **works** also WITHOUT this lines
**java.library.path=${localservicejarsPath};
spring.application.admin.enabled=true;**
and with fewer (only necessary for this call) jars to get the test method running

yours crashes as before with
(node:32436) UnhandledPromiseRejectionWarning: Error: java.lang.ClassNotFoundException: org.springframework.core.AliasRegistry

I think more tests couldn't I do, sorry

**Reply by tripower** [reply]
@MarkusJx 
for your information
have now made a test with a normal Java App 
means only a main class (springtest.TestStart) 
with creation of the object SystemInfoService and call of the method getComputerName
**=> works (at Eclipse)**
(with OpenJDK 11.0.7 x86 and localservice jars)

now also tested at Commandline 
java -cp C:\nodejs\electron-quick-start\localservice\*;springtest_onlyjar.jar springtest.TestStart
**=> works**

**Reply by MarkusJx** [reply]
The issue with win32 should be fixed now by #34, I've created a new issue for your problems with spring boot: #44, I'll start working on that in the following days

**Reply by tripower** [reply]
thx
if further informations needed just make a call
good luck and nice start into the week 🤗🍀

**Reply by tripower** [reply]
hello @MarkusJx 

have seen that you solved an other bug
will you do #44 before creating a new version?

greetings Klaus

**Reply by MarkusJx** [reply]
Yes, I'll resolve #38 and #44 before creating a new release

**Reply by tripower** [reply]
ok 
thx
good to know

greetings Klaus

**Reply by MarkusJx** [reply]
TL;DR: Spring boot uses its own class loader, you have to set the class path before jvm startup

So #38 is done, but I'll probably need some more time with the spring boot topic. My first guess was that I'm not correctly replacing the class loader, which may not be the case, as I wrote a test case for that, and it works: https://github.com/MarkusJx/node-java-bridge/blob/72078a40b38b29c528548dccf0c8b99e742f3b0a/test/ClassTest.test.ts#L143

I'm also setting the context class loader for each thread correctly, so that's not an issue. Thus, my only guess is that spring boot is using its own class loader which does not work well with me creating my own little class loading adventure. And it also makes sense they're using their own class loader as spring boot supports hot reload to some extend and you'd probably need your own class loader for that to work.

Yeah, looking at your logs, spring boot definitely uses its own class loader, so the only option to get this to work would be adding the classes at application startup. I'll probably add a startup option allowing you to set your classpath more conveniently (as this is btw also a thing `joeferner/java-bridge` does). But you'll have to make sure to call `ensureJvm` *before calling any other method* from this package, as any method call probably itself will call `ensureJvm` and calling that method multiple times will cause all calls after the first one to be no-ops, as I can't safely shut down/restart the jvm. This is a limitation mostly given by rust or rather its core principles as destroying the jvm would cause undefined behaviour, which is a thing that anyone should try to avoid. Yeah.

**Reply by tripower** [reply]
thanks
tests after a new build is available will show more
hope that was the solution

hope also that the jvm will only one time at the startup initialized (startup of spring classes)
otherwise it would be performance overkill

so fingers crossed that all works 🤗🙏
and #44 will also soon fixed. thx

greetings Klaus
